File: src/java/org/apache/fop/pdf/PDFTextUtil.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* $Id$ */

package org.apache.fop.pdf;

import java.awt.geom.AffineTransform;

/**
 * Utility class for generating PDF text objects. It needs to be subclassed to add writing
 * functionality (see {@link #write(String)}).
 */
public abstract class PDFTextUtil {

    /** The number of decimal places. */
    private static final int DEC = 8;

    /** PDF text rendering mode: Fill text */
    public static final int TR_FILL = 0;
    /** PDF text rendering mode: Stroke text */
    public static final int TR_STROKE = 1;
    /** PDF text rendering mode: Fill, then stroke text */
    public static final int TR_FILL_STROKE = 2;
    /** PDF text rendering mode: Neither fill nor stroke text (invisible) */
    public static final int TR_INVISIBLE = 3;
    /** PDF text rendering mode: Fill text and add to path for clipping */
    public static final int TR_FILL_CLIP = 4;
    /** PDF text rendering mode: Stroke text and add to path for clipping */
    public static final int TR_STROKE_CLIP = 5;
    /** PDF text rendering mode: Fill, then stroke text and add to path for clipping */
    public static final int TR_FILL_STROKE_CLIP = 6;
    /** PDF text rendering mode: Add text to path for clipping */
    public static final int TR_CLIP = 7;

    private boolean inTextObject;
    private String startText;
    private String endText;
    private boolean useMultiByte;
    private StringBuffer bufTJ;
    private int textRenderingMode = TR_FILL;

    private String currentFontName;
    private double currentFontSize;

    /**
     * Main constructor.
     */
    public PDFTextUtil() {
        //nop
    }

    /**
     * Writes PDF code.
     * @param code the PDF code to write
     */
    protected abstract void write(String code);

    /**
     * Writes PDF code.
     * @param code the PDF code to write
     */
    protected abstract void write(StringBuffer code);

    private void writeAffineTransform(AffineTransform at, StringBuffer sb) {
        double[] lt = new double[6];
        at.getMatrix(lt);
        PDFNumber.doubleOut(lt[0], DEC, sb);
        sb.append(' ');
        PDFNumber.doubleOut(lt[1], DEC, sb);
        sb.append(' ');
        PDFNumber.doubleOut(lt[2], DEC, sb);
        sb.append(' ');
        PDFNumber.doubleOut(lt[3], DEC, sb);
        sb.append(' ');
        PDFNumber.doubleOut(lt[4], DEC, sb);
        sb.append(' ');
        PDFNumber.doubleOut(lt[5], DEC, sb);
    }

    private static void writeChar(char ch, StringBuffer sb, boolean multibyte) {
        if (!multibyte) {
            if (ch < 32 || ch > 127) {
                sb.append("\\").append(Integer.toOctalString(ch));
            } else {
                switch (ch) {
                case '(':
                case ')':
                case '\\':
                    sb.append('\\');
                    break;
                default:
                }
                sb.append(ch);
            }
        } else {
            PDFText.toUnicodeHex(ch, sb);
        }
    }

    private void writeChar(char ch, StringBuffer sb) {
        writeChar(ch, sb, useMultiByte);
    }

    private void checkInTextObject() {
        if (!inTextObject) {
            throw new IllegalStateException("Not in text object");
        }
    }

    /**
     * Indicates whether we are in a text object or not.
     * @return true if we are in a text object
     */
    public boolean isInTextObject() {
        return inTextObject;
    }

    /**
     * Called when a new text object should be started. Be sure to call setFont() before
     * issuing any text painting commands.
     */
    public void beginTextObject() {
        if (inTextObject) {
            throw new IllegalStateException("Already in text object");
        }
        write("BT\n");
        this.inTextObject = true;
    }

    /**
     * Called when a text object should be ended.
     */
    public void endTextObject() {
        checkInTextObject();
        write("ET\n");
        this.inTextObject = false;
        initValues();
    }

    /**
     * Resets the state fields.
     */
    protected void initValues() {
        this.currentFontName = null;
        this.currentFontSize = 0.0;
        this.textRenderingMode = TR_FILL;
    }

    /**
     * Creates a "cm" command.
     * @param at the transformation matrix
     */
    public void concatMatrix(AffineTransform at) {
        if (!at.isIdentity()) {
            writeTJ();
            StringBuffer sb = new StringBuffer();
            writeAffineTransform(at, sb);
            sb.append(" cm\n");
            write(sb);
        }
    }

    /**
     * Writes a "Tf" command, setting a new current font.
     * @param fontName the name of the font to select
     * @param fontSize the font size (in points)
     */
    public void writeTf(String fontName, double fontSize) {
        checkInTextObject();
        StringBuffer sb = new StringBuffer();
        sb.append('/');
        sb.append(fontName);
        sb.append(' ');
        PDFNumber.doubleOut(fontSize, 6, sb);
        sb.append(" Tf\n");
        write(sb);
        this.startText = useMultiByte ? "<" : "(";
        this.endText = useMultiByte ? ">" : ")";
    }

    /**
     * Updates the current font. This method only writes a "Tf" if the current font changes.
     * @param fontName the name of the font to select
     * @param fontSize the font size (in points)
     * @param multiByte true indicates the font is a multi-byte font, false means single-byte
     */
    public void updateTf(String fontName, double fontSize, boolean multiByte) {
        checkInTextObject();
        if (!fontName.equals(this.currentFontName) || (fontSize != this.currentFontSize)) {
            writeTJ();
            this.currentFontName = fontName;
            this.currentFontSize = fontSize;
            this.useMultiByte = multiByte;
            writeTf(fontName, fontSize);
        }
    }

    /**
     * Sets the text rendering mode.
     * @param mode the rendering mode (value 0 to 7, see PDF Spec, constants: TR_*)
     */
    public void setTextRenderingMode(int mode) {
        if (mode < 0 || mode > 7) {
            throw new IllegalArgumentException(
                    "Illegal value for text rendering mode. Expected: 0-7");
        }
        if (mode != this.textRenderingMode) {
            writeTJ();
            this.textRenderingMode = mode;
            write(this.textRenderingMode + " Tr\n");
        }
    }

    /**
     * Sets the text rendering mode.
     * @param fill true if the text should be filled
     * @param stroke true if the text should be stroked
     * @param addToClip true if the path should be added for clipping
     */
    public void setTextRenderingMode(boolean fill, boolean stroke, boolean addToClip) {
        int mode;
        if (fill) {
            mode = (stroke ? 2 : 0);
        } else {
            mode = (stroke ? 1 : 3);
        }
        if (addToClip) {
            mode += 4;
        }
        setTextRenderingMode(mode);
    }

    /**
     * Writes a "Tm" command, setting a new text transformation matrix.
     * @param localTransform the new text transformation matrix
     */
    public void writeTextMatrix(AffineTransform localTransform) {
        StringBuffer sb = new StringBuffer();
        writeAffineTransform(localTransform, sb);
        sb.append(" Tm ");
        write(sb);
    }

    /**
     * Writes a char to the "TJ-Buffer".
     * @param codepoint the mapped character (code point/character code)
     */
    public void writeTJMappedChar(char codepoint) {
        if (bufTJ == null) {
            bufTJ = new StringBuffer();
        }
        if (bufTJ.length() == 0) {
            bufTJ.append('[');
            bufTJ.append(startText);
        }
        writeChar(codepoint, bufTJ);
    }

    /**
     * Writes a glyph adjust value to the "TJ-Buffer".

     * <p>Assumes the following:</p>
     * <ol>
     * <li>if buffer is currently empty, then this is the start of the array object
     * that encodes the adjustment and character values, and, therfore, a LEFT
     * SQUARE BRACKET '[' must be prepended; and
     * </li>
     * <li>otherwise (the buffer is
     * not empty), then the last element written to the buffer was a mapped
     * character, and, therefore, a terminating '&gt;' or ')' followed by a space
     * must be appended to the buffer prior to appending the adjustment value.
     * </li>
     * </ol>
     * @param adjust the glyph adjust value in thousands of text unit space.
     */
    public void adjustGlyphTJ(double adjust) {
        if (bufTJ == null) {
            bufTJ = new StringBuffer();
        }
        if (bufTJ.length() == 0) {
            bufTJ.append('[');
        } else {
            bufTJ.append(endText);
            bufTJ.append(' ');
        }
        PDFNumber.doubleOut(adjust, DEC - 4, bufTJ);
        bufTJ.append(' ');
        bufTJ.append(startText);
    }

    /**
     * Writes a "TJ" command, writing out the accumulated buffer with the characters and glyph
     * positioning values. The buffer is reset afterwards.
     */
    public void writeTJ() {
        if (isInString()) {
            bufTJ.append(endText);
            bufTJ.append("] TJ\n");
            write(bufTJ);
            bufTJ.setLength(0);
        }
    }

    private boolean isInString() {
        return bufTJ != null && bufTJ.length() > 0;
    }

    /**
     * Writes a "Td" command with specified x and y coordinates.
     * @param x coordinate
     * @param y coordinate
     */
    public void writeTd(double x, double y) {
        StringBuffer sb = new StringBuffer();
        PDFNumber.doubleOut(x, DEC, sb);
        sb.append(' ');
        PDFNumber.doubleOut(y, DEC, sb);
        sb.append(" Td\n");
        write(sb);
    }

    /**
     * Writes a "Tj" command with specified character code.
     * @param ch character code to write
     */
    public void writeTj(char ch) {
        StringBuffer sb = new StringBuffer();
        sb.append('<');
        writeChar(ch, sb, true);
        sb.append('>');
        sb.append(" Tj\n");
        write(sb);
    }

}

File: dotnet/src/Fop.Pdf/PDFTextUtil.cs
using System.Globalization;
using System.Text;

namespace Fop.Pdf;

/// <summary>
/// Utility for generating PDF text objects. Subclass it and implement <see cref="Write(string)"/>
/// / <see cref="Write(StringBuilder)"/> to add writing functionality.
/// </summary>
public abstract class PDFTextUtil
{
    /// <summary>The number of decimal places used for coordinate output.</summary>
    private const int Dec = 8;

    /// <summary>PDF text rendering mode: Fill text.</summary>
    public const int TrFill = 0;
    /// <summary>PDF text rendering mode: Stroke text.</summary>
    public const int TrStroke = 1;
    /// <summary>PDF text rendering mode: Fill, then stroke text.</summary>
    public const int TrFillStroke = 2;
    /// <summary>PDF text rendering mode: Neither fill nor stroke text (invisible).</summary>
    public const int TrInvisible = 3;
    /// <summary>PDF text rendering mode: Fill text and add to path for clipping.</summary>
    public const int TrFillClip = 4;
    /// <summary>PDF text rendering mode: Stroke text and add to path for clipping.</summary>
    public const int TrStrokeClip = 5;
    /// <summary>PDF text rendering mode: Fill, then stroke text and add to path for clipping.</summary>
    public const int TrFillStrokeClip = 6;
    /// <summary>PDF text rendering mode: Add text to path for clipping.</summary>
    public const int TrClip = 7;

    private bool _inTextObject;
    private string? _startText;
    private string? _endText;
    private bool _useMultiByte;
    private bool _useCid;
    private StringBuilder? _bufTJ;
    private int _textRenderingMode = TrFill;

    private string? _currentFontName;
    private double _currentFontSize;

    /// <summary>Writes PDF code.</summary>
    protected abstract void Write(string code);

    /// <summary>Writes PDF code.</summary>
    protected abstract void Write(StringBuilder code);

    private static void WriteAffineTransform(PdfAffineTransform at, StringBuilder sb)
    {
        double[] lt = at.GetMatrix();
        PDFNumber.DoubleOut(lt[0], Dec, sb);
        sb.Append(' ');
        PDFNumber.DoubleOut(lt[1], Dec, sb);
        sb.Append(' ');
        PDFNumber.DoubleOut(lt[2], Dec, sb);
        sb.Append(' ');
        PDFNumber.DoubleOut(lt[3], Dec, sb);
        sb.Append(' ');
        PDFNumber.DoubleOut(lt[4], Dec, sb);
        sb.Append(' ');
        PDFNumber.DoubleOut(lt[5], Dec, sb);
    }

    private static void WriteChar(int codePoint, StringBuilder sb, bool multibyte, bool cid)
    {
        if (!multibyte)
        {
            if (cid || codePoint < 32 || codePoint > 127)
            {
                sb.Append('\\').Append(ToOctal(codePoint));
            }
            else
            {
                switch (codePoint)
                {
                    case '(':
                    case ')':
                    case '\\':
                        sb.Append('\\');
                        break;
                    default:
                        break;
                }
                sb.Append(char.ConvertFromUtf32(codePoint));
            }
        }
        else
        {
            PDFText.ToUnicodeHex(codePoint, sb);
        }
    }

    private void WriteChar(int codePoint, StringBuilder sb) =>
        WriteChar(codePoint, sb, _useMultiByte, _useCid);

    private void CheckInTextObject()
    {
        if (!_inTextObject)
        {
            throw new InvalidOperationException("Not in text object");
        }
    }

    /// <summary>Indicates whether we are in a text object.</summary>
    public bool IsInTextObject => _inTextObject;

    /// <summary>
    /// Begins a new text object. Be sure to call <see cref="WriteTf"/> before issuing any text
    /// painting commands.
    /// </summary>
    public void BeginTextObject()
    {
        if (_inTextObject)
        {
            throw new InvalidOperationException("Already in text object");
        }
        Write("BT\n");
        _inTextObject = true;
    }

    /// <summary>Ends the current text object.</summary>
    public void EndTextObject()
    {
        CheckInTextObject();
        Write("ET\n");
        _inTextObject = false;
        InitValues();
    }

    /// <summary>Resets the state fields.</summary>
    protected virtual void InitValues()
    {
        _currentFontName = null;
        _currentFontSize = 0.0;
        _textRenderingMode = TrFill;
    }

    /// <summary>Creates a "cm" command from the transformation matrix.</summary>
    public void ConcatMatrix(PdfAffineTransform at)
    {
        if (!at.IsIdentity)
        {
            WriteTJ();
            var sb = new StringBuilder();
            WriteAffineTransform(at, sb);
            sb.Append(" cm\n");
            Write(sb);
        }
    }

    /// <summary>Writes a "Tf" command, setting a new current font.</summary>
    public void WriteTf(string fontName, double fontSize)
    {
        CheckInTextObject();
        var sb = new StringBuilder();
        sb.Append('/');
        sb.Append(fontName);
        sb.Append(' ');
        PDFNumber.DoubleOut(fontSize, 6, sb);
        sb.Append(" Tf\n");
        Write(sb);
        _startText = _useMultiByte ? "<" : "(";
        _endText = _useMultiByte ? ">" : ")";
    }

    /// <summary>Updates the current font, writing a "Tf" only if it changed.</summary>
    public void UpdateTf(string fontName, double fontSize, bool multiByte, bool cid)
    {
        CheckInTextObject();
        if (fontName != _currentFontName || fontSize != _currentFontSize)
        {
            WriteTJ();
            _currentFontName = fontName;
            _currentFontSize = fontSize;
            _useMultiByte = multiByte;
            _useCid = cid;
            WriteTf(fontName, fontSize);
        }
    }

    /// <summary>Sets the text rendering mode (0..7, see the TR_* constants).</summary>
    public void SetTextRenderingMode(int mode)
    {
        if (mode is < 0 or > 7)
        {
            throw new ArgumentException("Illegal value for text rendering mode. Expected: 0-7");
        }
        if (mode != _textRenderingMode)
        {
            WriteTJ();
            _textRenderingMode = mode;
            Write(_textRenderingMode.ToString(CultureInfo.InvariantCulture) + " Tr\n");
        }
    }

    /// <summary>Sets the text rendering mode from the fill/stroke/clip flags.</summary>
    public void SetTextRenderingMode(bool fill, bool stroke, bool addToClip)
    {
        int mode = fill ? (stroke ? 2 : 0) : (stroke ? 1 : 3);
        if (addToClip)
        {
            mode += 4;
        }
        SetTextRenderingMode(mode);
    }

    /// <summary>Writes a "Tm" command, setting a new text transformation matrix.</summary>
    public void WriteTextMatrix(PdfAffineTransform localTransform)
    {
        var sb = new StringBuilder();
        WriteAffineTransform(localTransform, sb);
        sb.Append(" Tm ");
        Write(sb);
    }

    /// <summary>Writes a mapped char to the "TJ" buffer.</summary>
    public void WriteTJMappedChar(char ch) => WriteTJMappedCodePoint(ch);

    /// <summary>Writes a mapped code point to the "TJ" buffer.</summary>
    public void WriteTJMappedCodePoint(int codePoint)
    {
        _bufTJ ??= new StringBuilder();
        if (_bufTJ.Length == 0)
        {
            _bufTJ.Append('[');
            _bufTJ.Append(_startText);
        }
        WriteChar(codePoint, _bufTJ);
    }

    /// <summary>Writes a glyph adjust value (in thousands of text unit space) to the "TJ" buffer.</summary>
    public void AdjustGlyphTJ(double adjust)
    {
        _bufTJ ??= new StringBuilder();
        if (_bufTJ.Length == 0)
        {
            _bufTJ.Append('[');
        }
        else
        {
            _bufTJ.Append(_endText);
            _bufTJ.Append(' ');
        }
        PDFNumber.DoubleOut(adjust, Dec - 4, _bufTJ);
        _bufTJ.Append(' ');
        _bufTJ.Append(_startText);
    }

    /// <summary>
    /// Writes a "TJ" command, flushing the accumulated buffer of characters and positioning
    /// values. The buffer is reset afterwards.
    /// </summary>
    public void WriteTJ()
    {
        if (IsInString())
        {
            _bufTJ!.Append(_endText);
            _bufTJ.Append("] TJ\n");
            Write(_bufTJ);
            _bufTJ.Length = 0;
        }
    }

    private bool IsInString() => _bufTJ is { Length: > 0 };

    /// <summary>Writes a "Td" command with the specified x and y coordinates.</summary>
    public void WriteTd(double x, double y)
    {
        var sb = new StringBuilder();
        PDFNumber.DoubleOut(x, Dec, sb);
        sb.Append(' ');
        PDFNumber.DoubleOut(y, Dec, sb);
        sb.Append(" Td\n");
        Write(sb);
    }

    /// <summary>Writes a "Tj" command with the specified character code.</summary>
    public void WriteTj(int ch, bool multibyte, bool cid)
    {
        var sb = new StringBuilder();
        sb.Append(_startText);
        WriteChar(ch, sb, multibyte, cid);
        sb.Append(_endText);
        sb.Append(" Tj\n");
        Write(sb);
    }

    // Java Integer.toOctalString: unsigned octal, no leading zeros.
    private static string ToOctal(int value)
    {
        if (value == 0)
        {
            return "0";
        }
        uint u = unchecked((uint)value);
        var sb = new StringBuilder();
        while (u != 0)
        {
            sb.Insert(0, (char)('0' + (int)(u & 7)));
            u >>= 3;
        }
        return sb.ToString();
    }
}
